File: src/2024/day-01/typescript/part-2.ts
const words = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine",];
export function Part2(input: string) {
  return input
    .trim()
    .split("\n")
    .map((line) => line.trim().split("")
      .map((char, index) =>
        char.match(/\d/)
          ? char
          : words.findIndex((word) => line.slice(index).startsWith(word)) + 1,
      )
      .filter((num) => num !== 0)
      .map(Number))
    .map((numbers) => numbers[0] * 10 + numbers[numbers.length - 1])
    .reduce((a, b) => a + b);
}

File: src/2024/day-01/python/part_2.py
import re

words = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]


# Word to number helper function
def w2n(value: str) -> int:
    if value.isnumeric():
        return int(value)
    return words.index(value) + 1


def Part2(input_str):
    lines = input_str.strip().split("\n")
    # Pattern to match a number word
    pattern = "one|two|three|four|five|six|seven|eight|nine"
    # Reverse the pattern so that we can search backwards from end
    reverse_pattern = pattern[::-1]
    total = 0
    for line in [ln.strip() for ln in lines]:
        # Find the first digit
        first = re.search(rf"(\d|{pattern})", line)
        # Find the last digit by searching for reversed pattern
        # on the reversed string
        last = re.search(rf"(\d|{reverse_pattern})", line[::-1])
        # Update running total
        total += w2n(first.group()) * 10 + w2n(last.group()[::-1])
    return total
